src/services/communication/spi.py

```diff
@@ -0,0 +1,146 @@
+    select the slave. The commands are this:
+    - SLAVE_SELECT: this command is sended under the number '1'
+    - SLAVE_DESELECT: this command is sended under the number '2'
+    - NORMAL_PACKET: this command is sended under the number '0',
+        followed by the message intented to send.
+    To the other side, if you are a slave, you must check the command before process the data.
+    Communication itself has been implemented using UDP protocol.
+    Messages are handled using queues and threads, one for transmission and other for reception.
+    '''
+    @abc.abstractmethod
+    def _send(self): pass
+    
+    @abc.abstractmethod
+    def _recv(self): pass
+
+    def __init__(self, ip, port): 
+        self.socket = socket.socket(type=socket.SOCK_DGRAM) # Socket used to send and receive SPI data
+        self.socket.bind((ip, port))
+
+        self.transmission_queue = queue.Queue() # Queue used to store messages that should be sended
+        self.reception_queue = queue.Queue() # Queue used to store received messages
+        self.transmission_thread = Thread(target=self._send , args=("Transmitter")) # Thread that manages data transmission
+        self.reception_thread = Thread(target=self._recv , args=("Receiver")) # Thread that manages data reception
+
+        self.transmission_thread.start()
+        self.reception_thread.start()
+
+    def receive(self) -> bytes: 
+        return self.reception_queue.get
+    
+    @abc.abstractmethod
+    def transmit(self): pass
+
+    def __del__(self):
+        self.socket.close()
+        self.stop = True
+        self.transmission_thread.join()
+        self.reception_thread.join()
+        
+
+class SPIMaster(SPIPeripheral): 
+    def _send(self):
+        while(not self.stop):
+            msg = self.transmission_queue.get
+            total_sent = 0
+            while total_sent < len(msg):
+                sent = self.socket.sendto(msg[sent:], self.send_address)
+                if sent == 0:
+                    raise RuntimeError("Socket connection broken")
+                total_sent += sent
+
+    def _recv(self):
+        while(not self.stop):
+            received = self.socket.recv(1024)
+            self.reception_queue.put(received)
+            
+    def transmit(self, msg, ip_slave, port_slave):
+        '''
+        Transmits data simulating SPI communication, considering that you are a Master.
+
+        Args:
+            msg: message to be sended
+            ip_slave: ip address of the slave you want to select and send the data
+            port_slave: port of the slave you want to select and send the data
+        '''
+        if (self.send_address != (ip_slave, port_slave)):
+            self.transmission_queue.put(b'2') # Sending SLAVE_SELECT command
+            self.send_address = (ip_slave, port_slave)
+            self.transmission_queue.put(b'1') # Sending SLAVE_DESELECT command
+        self.transmission_queue.put(b'0' + msg)
```

Comment:
I think oscar is right, you are selecting, deselecting and then sending, you should instead select, send and deselect

Reply:
Sorry, the comments was exchanged, and that's probably cause confusion. I'm sending SLAVE_DESELECT to the OLD slave, then change the slave addres to the NEW slave, and then send SLAVE_SELECT to the NEW slave